Mage.Sets/src/mage/cards/c/ChandraHeartOfFire.java

```diff
@@ -0,0 +1,135 @@
+        damageAbility.addTarget(new TargetAnyTarget());
+        this.addAbility(damageAbility);
+
+        // −9: Search your graveyard and library for any number of red instant and/or sorcery spells, exile them, then shuffle your library. You may cast them this turn. Add six {R}.
+        Ability ultimateAbility = new LoyaltyAbility(new ChandraHeartOfFireUltimateEffect(), -9);
+        ultimateAbility.addEffect(new BasicManaEffect(Mana.RedMana(6)).setText("Add six {R}"));
+        this.addAbility(ultimateAbility);
+    }
+
+    private ChandraHeartOfFire(final ChandraHeartOfFire card) {
+        super(card);
+    }
+
+    @Override
+    public ChandraHeartOfFire copy() {
+        return new ChandraHeartOfFire(this);
+    }
+}
+
+class ChandraHeartOfFireUltimateEffect extends OneShotEffect {
+
+    private static final FilterCard filter = new FilterCard();
+
+    static {
+        filter.add(new ColorPredicate(ObjectColor.RED));
+        filter.add(Predicates.or(CardType.INSTANT.getPredicate(), CardType.SORCERY.getPredicate()));
+    }
+
+    ChandraHeartOfFireUltimateEffect() {
+        super(Outcome.Benefit);
+        staticText = "Search your graveyard and library for any number of red instant and/or sorcery spells, exile them, then shuffle your library. You may cast them this turn";
+    }
+
+    private ChandraHeartOfFireUltimateEffect(ChandraHeartOfFireUltimateEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public ChandraHeartOfFireUltimateEffect copy() {
+        return new ChandraHeartOfFireUltimateEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            Set<Card> exiledCards = new HashSet<>();
+
+            filter.setMessage("red instant or sorcery in your graveyard");
+            TargetCard target = new TargetCard(0, Integer.MAX_VALUE, Zone.GRAVEYARD, filter);
+            if (controller.choose(Outcome.Exile, controller.getGraveyard(), target, game)) {
+                List<UUID> targets = target.getTargets();
+                controller.moveCards(new CardsImpl(targets), Zone.EXILED, source, game);
+                exiledCards.addAll(targets.stream().map(game::getCard).collect(Collectors.toList()));
+            }
+
+            filter.setMessage("red instant or sorcery in your library");
+            Cards cardsInLibrary = new CardsImpl();
+            cardsInLibrary.addAll(controller.getLibrary().getCards(game));
+            target = new TargetCardInLibrary(0, Integer.MAX_VALUE, filter);
+            if (controller.choose(Outcome.Exile, cardsInLibrary, target, game)) {
+                List<UUID> targets = target.getTargets();
+                controller.moveCards(new CardsImpl(targets), Zone.EXILED, source, game);
+                exiledCards.addAll(targets.stream().map(game::getCard).collect(Collectors.toList()));
```

Comment:
Missing `null` check after `game::getCard`. All stream code can be replaced by more simpler code (get cards from targets):
* `Set<Card> cards = new CardsImpl(target.getTargets()).getCards(game);`